src/AndersonAcceleration.f90

```diff
@@ -196,59 +143,217 @@ SUBROUTINE init_AndersonAccelerationType(solver,MPIEnv,Params)
 !> @brief Clears the Anderson Acceleration Type
 !> @param solver The solver to act on
 !>
-!> This routine clears the data spaces
-!>
 SUBROUTINE clear_AndersonAccelerationType(solver)
   CLASS(AndersonAccelerationType),INTENT(INOUT) :: solver
 
-  NULLIFY(solver%MPIparallelEnv)
-  IF(solver%OMPparallelEnv%isInit()) CALL solver%OMPparallelEnv%clear
-  solver%iter=0
-  solver%n=-1
-  solver%depth=-1
-  solver%beta=0.0_SRK
-  IF(ASSOCIATED(solver%X)) THEN
-    CALL solver%X%clear()
-    DEALLOCATE(solver%X)
-    NULLIFY(solver%X)
+  IF(solver%isInit) THEN
+    NULLIFY(solver%MPIparallelEnv)
+    solver%s=0
+    solver%N=-1
+    solver%depth=-1
+    solver%start=1
+    solver%depth=-1
+    solver%beta=0.0_SRK
+    DEALLOCATE(solver%x)
+    DEALLOCATE(solver%Gx)
+    DEALLOCATE(solver%r)
+    DEALLOCATE(solver%A)
+    DEALLOCATE(solver%alpha)
+    DEALLOCATE(solver%b)
+    solver%isInit=.FALSE.
   ENDIF
-#ifdef FUTILITY_HAVE_Trilinos_NOX
-  CALL Anderson_Destroy(solver%id)
-#endif
-  solver%isInit=.FALSE.
+
 ENDSUBROUTINE clear_AndersonAccelerationType
 !
 !-------------------------------------------------------------------------------
-!> @brief Single step of the Anderson acceleration
-!> @param solver The anderson solver to act on
+!> @brief Single step of Anderson acceleration acting on solution vectors
+!> @param solver Anderson solver to take step with
+!> @param New_Gx new solution iterate
 !>
 !> This routine takes a single anderson acceleration step
 !>
-SUBROUTINE step_AndersonAccelerationType(solver)
+FUNCTION step_AndersonAccelerationType(solver,New_Gx) RESULT(x_tmp)
   CLASS(AndersonAccelerationType),INTENT(INOUT) :: solver
-#ifdef FUTILITY_HAVE_Trilinos_NOX
-  IF(solver%iter==0) THEN
-    CALL Anderson_Reset(solver%id)
-  ELSE
-    CALL Anderson_Update(solver%id)
+  REAL(SRK),INTENT(IN) :: New_Gx(:)
+
+  INTEGER(SIK) :: i,j,k,depth_s
+  REAL(SRK) :: x_tmp(solver%N)
+
+  !Update iteration counter
+  solver%s=solver%s+1
+
+  IF(solver%s >= solver%start) THEN
+    depth_s=MIN(solver%depth,solver%s-solver%start)
+    !Push back vectors to make room for new ones
+    DO i=depth_s,1,-1
+      solver%Gx(:,i+1)=solver%Gx(:,i)
+      solver%r(:,i+1)=solver%r(:,i)
+    ENDDO
+
+    !Set new vectors based on input
+    solver%Gx(:,1)=New_Gx(:)
+    solver%r(:,1)=solver%Gx(:,1)-solver%x(:,1)
+
+    !Get fit coefficients
+    IF(depth_s == 1) THEN
+      solver%A(1,1)=0.0_SRK
+      solver%b(1)=0.0_SRK
+      DO j=1,solver%N
```

Comment:
Are these standard norms?  If you do this with vector types, this could be generic for any vector type and use the built in norms.

Reply:
this would make x, Gx, and r an array pointers to vector types

Reply:
Not really any norms going on here. Just sums of differences.